File: core/regression_engine.py
"""
OmniCalc Pro Regression & Curve Fitting Engine (Python Desktop Parity - Pure Python)
Supports Linear, Polynomial, Exponential, Logarithmic, and Power regressions.
"""

import math
from typing import List, Dict, Any, Tuple


class RegressionEngine:
    @staticmethod
    def linear_fit(x_pts: List[float], y_pts: List[float]) -> Dict[str, Any]:
        n = len(x_pts)
        if n < 2 or len(y_pts) != n:
            raise ValueError("Linear regression requires at least 2 points with equal x and y lengths.")

        sum_x = sum(x_pts)
        sum_y = sum(y_pts)
        sum_xy = sum(x * y for x, y in zip(x_pts, y_pts))
        sum_x2 = sum(x * x for x in x_pts)

        denom = n * sum_x2 - sum_x * sum_x
        if abs(denom) < 1e-12:
            raise ValueError("Vertical line or degenerate dataset: x values are constant.")

        slope = (n * sum_xy - sum_x * sum_y) / denom
        intercept = (sum_y - slope * sum_x) / n

        mean_y = sum_y / n
        ss_tot = sum((y - mean_y) ** 2 for y in y_pts)
        ss_res = sum((y - (slope * x + intercept)) ** 2 for x, y in zip(x_pts, y_pts))

        r2 = 1.0 - (ss_res / ss_tot) if ss_tot > 0 else 1.0

        sign_str = '+' if intercept >= 0 else '-'
        eq_str = f"y = {slope:.4f}x {sign_str} {abs(intercept):.4f}"

        return {
            "type": "linear",
            "slope": float(slope),
            "intercept": float(intercept),
            "r2": float(r2),
            "equation": eq_str,
            "ss_res": float(ss_res),
            "ss_tot": float(ss_tot),
        }

    @staticmethod
    def polynomial_fit(x_pts: List[float], y_pts: List[float], degree: int = 2) -> Dict[str, Any]:
        n = len(x_pts)
        if n <= degree or len(y_pts) != n:
            raise ValueError(f"Polynomial regression of degree {degree} requires at least {degree + 1} points.")

        # Solve Normal Equations (X^T X) C = X^T Y via Gaussian Elimination
        m = degree + 1
        A = [[0.0] * m for _ in range(m)]
        B = [0.0] * m

        for i in range(m):
            for j in range(m):
                A[i][j] = sum((x ** (i + j)) for x in x_pts)
            B[i] = sum((y * (x ** i)) for x, y in zip(x_pts, y_pts))

        # Gaussian Elimination with Partial Pivoting
        for i in range(m):
            max_row = i
            for k in range(i + 1, m):
                if abs(A[k][i]) > abs(A[max_row][i]):
                    max_row = k
            A[i], A[max_row] = A[max_row], A[i]
            B[i], B[max_row] = B[max_row], B[i]

            pivot = A[i][i]
            if abs(pivot) < 1e-12:
                raise ValueError("Degenerate matrix in polynomial fitting.")

            for k in range(i + 1, m):
                factor = A[k][i] / pivot
                for j in range(i, m):
                    A[k][j] -= factor * A[i][j]
                B[k] -= factor * B[i]

        # Back Substitution
        coeffs_asc = [0.0] * m
        for i in range(m - 1, -1, -1):
            s = B[i] - sum(A[i][j] * coeffs_asc[j] for j in range(i + 1, m))
            coeffs_asc[i] = s / A[i][i]

        # Reverse to descending order (highest degree first)
        coeffs = coeffs_asc[::-1]

        def poly_eval(x_val: float) -> float:
            res = 0.0
            for c in coeffs:
                res = res * x_val + c
            return res

        mean_y = sum(y_pts) / n
        ss_tot = sum((y - mean_y) ** 2 for y in y_pts)
        ss_res = sum((y - poly_eval(x)) ** 2 for x, y in zip(x_pts, y_pts))
        r2 = 1.0 - (ss_res / ss_tot) if ss_tot > 0 else 1.0

        # Build readable equation string
        terms = []
        deg = degree
        for c in coeffs:
            if deg == 0:
                terms.append(f"{c:.4f}")
            elif deg == 1:
                terms.append(f"{c:.4f}x")
            else:
                terms.append(f"{c:.4f}x^{deg}")
            deg -= 1
        eq_str = "y = " + " + ".join(terms).replace("+ -", "- ")

        return {
            "type": f"polynomial_deg_{degree}",
            "coefficients": [float(c) for c in coeffs],
            "r2": float(r2),
            "equation": eq_str,
            "ss_res": float(ss_res),
            "ss_tot": float(ss_tot),
        }

    @staticmethod
    def exponential_fit(x_pts: List[float], y_pts: List[float]) -> Dict[str, Any]:
        if any(y <= 0 for y in y_pts):
            raise ValueError("Exponential regression y = a * e^(b*x) requires all y > 0.")
        ln_y = [math.log(y) for y in y_pts]
        lin_res = RegressionEngine.linear_fit(x_pts, ln_y)
        b = lin_res["slope"]
        a = math.exp(lin_res["intercept"])

        mean_y = sum(y_pts) / len(y_pts)
        ss_tot = sum((y - mean_y) ** 2 for y in y_pts)
        ss_res = sum((y - a * math.exp(b * x)) ** 2 for x, y in zip(x_pts, y_pts))
        r2 = 1.0 - (ss_res / ss_tot) if ss_tot > 0 else 1.0

        return {
            "type": "exponential",
            "a": float(a),
            "b": float(b),
            "r2": float(r2),
            "equation": f"y = {a:.4f} * e^({b:.4f}x)",
            "ss_res": float(ss_res),
            "ss_tot": float(ss_tot),
        }

    @staticmethod
    def logarithmic_fit(x_pts: List[float], y_pts: List[float]) -> Dict[str, Any]:
        if any(x <= 0 for x in x_pts):
            raise ValueError("Logarithmic regression y = a + b * ln(x) requires all x > 0.")
        ln_x = [math.log(x) for x in x_pts]
        lin_res = RegressionEngine.linear_fit(ln_x, y_pts)
        b = lin_res["slope"]
        a = lin_res["intercept"]

        mean_y = sum(y_pts) / len(y_pts)
        ss_tot = sum((y - mean_y) ** 2 for y in y_pts)
        ss_res = sum((y - (a + b * math.log(x))) ** 2 for x, y in zip(x_pts, y_pts))
        r2 = 1.0 - (ss_res / ss_tot) if ss_tot > 0 else 1.0

        sign_str = '+' if b >= 0 else '-'
        return {
            "type": "logarithmic",
            "a": float(a),
            "b": float(b),
            "r2": float(r2),
            "equation": f"y = {a:.4f} {sign_str} {abs(b):.4f} * ln(x)",
            "ss_res": float(ss_res),
            "ss_tot": float(ss_tot),
        }

    @staticmethod
    def power_fit(x_pts: List[float], y_pts: List[float]) -> Dict[str, Any]:
        if any(x <= 0 for x in x_pts) or any(y <= 0 for y in y_pts):
            raise ValueError("Power regression y = a * x^b requires all x > 0 and y > 0.")
        ln_x = [math.log(x) for x in x_pts]
        ln_y = [math.log(y) for y in y_pts]
        lin_res = RegressionEngine.linear_fit(ln_x, ln_y)
        b = lin_res["slope"]
        a = math.exp(lin_res["intercept"])

        mean_y = sum(y_pts) / len(y_pts)
        ss_tot = sum((y - mean_y) ** 2 for y in y_pts)
        ss_res = sum((y - a * (x ** b)) ** 2 for x, y in zip(x_pts, y_pts))
        r2 = 1.0 - (ss_res / ss_tot) if ss_tot > 0 else 1.0

        return {
            "type": "power",
            "a": float(a),
            "b": float(b),
            "r2": float(r2),
            "equation": f"y = {a:.4f} * x^({b:.4f})",
            "ss_res": float(ss_res),
            "ss_tot": float(ss_tot),
        }

    @staticmethod
    def predict(fit_res: Dict[str, Any], x_val: float) -> float:
        model_type = fit_res.get("type", "")
        if model_type == "linear":
            return fit_res["slope"] * x_val + fit_res["intercept"]
        elif model_type.startswith("polynomial"):
            coeffs = fit_res["coefficients"]
            res = 0.0
            for c in coeffs:
                res = res * x_val + c
            return res
        elif model_type == "exponential":
            return fit_res["a"] * math.exp(fit_res["b"] * x_val)
        elif model_type == "logarithmic":
            if x_val <= 0:
                raise ValueError("x must be > 0 for logarithmic prediction.")
            return fit_res["a"] + fit_res["b"] * math.log(x_val)
        elif model_type == "power":
            if x_val <= 0:
                raise ValueError("x must be > 0 for power prediction.")
            return fit_res["a"] * (x_val ** fit_res["b"])
        else:
            raise ValueError(f"Unknown regression model type: {model_type}")

    @staticmethod
    def fit(x_pts: List[float], y_pts: List[float], model_type: str = "linear", degree: int = 2) -> Dict[str, Any]:
        model = model_type.lower()
        if model == "linear":
            return RegressionEngine.linear_fit(x_pts, y_pts)
        elif model == "polynomial":
            return RegressionEngine.polynomial_fit(x_pts, y_pts, degree=degree)
        elif model == "exponential":
            return RegressionEngine.exponential_fit(x_pts, y_pts)
        elif model == "logarithmic":
            return RegressionEngine.logarithmic_fit(x_pts, y_pts)
        elif model == "power":
            return RegressionEngine.power_fit(x_pts, y_pts)
        else:
            raise ValueError(f"Unsupported regression model type: {model_type}")

File: src/utils/regressionEngine.ts
/**
 * Regression & Curve Fitting Engine for OmniCalc Pro.
 * Computes Ordinary Least Squares (OLS) Linear, Polynomial (Quadratic & Cubic),
 * Exponential, Power, and Logarithmic regressions with complete ANOVA metrics (R², RSS, TSS, residuals).
 */

import { invertMatrix, multiplyMatrices, Matrix } from './matrix';

export interface DataPoint {
  x: number;
  y: number;
}

export interface ResidualPoint {
  x: number;
  yObserved: number;
  yPredicted: number;
  residual: number;
}

export interface RegressionResult {
  type: 'linear' | 'polynomial' | 'exponential' | 'power' | 'logarithmic';
  equation: string;
  r: number; // Pearson correlation coefficient
  rSquared: number; // Coefficient of determination
  coefficients: number[];
  predict: (x: number) => number;
  residuals: ResidualPoint[];
  standardError: number;
  degree?: number;
}

/**
 * Computes Linear Regression: y = mx + b.
 */
export function fitLinearRegression(points: DataPoint[]): RegressionResult | null {
  const n = points.length;
  if (n < 2) return null;

  let sumX = 0;
  let sumY = 0;
  let sumXY = 0;
  let sumX2 = 0;

  for (const p of points) {
    sumX += p.x;
    sumY += p.y;
    sumXY += p.x * p.y;
    sumX2 += p.x * p.x;
  }

  const denom = n * sumX2 - sumX * sumX;
  if (Math.abs(denom) < 1e-12) return null; // Vertical line

  const slope = (n * sumXY - sumX * sumY) / denom;
  const intercept = (sumY - slope * sumX) / n;

  // Correlation r and R²
  const meanY = sumY / n;
  let ssTot = 0;
  let ssRes = 0;

  const residuals: ResidualPoint[] = [];
  for (const p of points) {
    const yPred = slope * p.x + intercept;
    const res = p.y - yPred;
    ssTot += (p.y - meanY) * (p.y - meanY);
    ssRes += res * res;
    residuals.push({
      x: p.x,
      yObserved: p.y,
      yPredicted: yPred,
      residual: res,
    });
  }

  const rSquared = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 1;
  const r = (slope >= 0 ? 1 : -1) * Math.sqrt(rSquared);
  const standardError = n > 2 ? Math.sqrt(ssRes / (n - 2)) : 0;

  const sign = intercept >= 0 ? '+' : '-';
  const eq = `y = ${slope.toFixed(4)}x ${sign} ${Math.abs(intercept).toFixed(4)}`;

  return {
    type: 'linear',
    equation: eq,
    r,
    rSquared,
    coefficients: [intercept, slope],
    predict: (x: number) => slope * x + intercept,
    residuals,
    standardError,
  };
}

/**
 * Computes Polynomial Regression of specified degree (e.g. 2 for quadratic, 3 for cubic)
 * using the OLS normal equation: (X^T X) * Beta = X^T Y.
 */
export function fitPolynomialRegression(points: DataPoint[], degree: number = 2): RegressionResult | null {
  const n = points.length;
  const numCoeffs = degree + 1;
  if (n < numCoeffs || degree < 1) return null;

  // Build design matrix X: n x numCoeffs where X[i][j] = x_i^j
  const X: Matrix = Array.from({ length: n }, (_, i) =>
    Array.from({ length: numCoeffs }, (_, j) => Math.pow(points[i].x, j))
  );

  // Y vector: n x 1
  const Y: Matrix = points.map((p) => [p.y]);

  // Compute X^T
  const XT: Matrix = Array.from({ length: numCoeffs }, (_, j) =>
    Array.from({ length: n }, (_, i) => X[i][j])
  );

  // Normal equations: (X^T * X) * Beta = X^T * Y
  const XTX = multiplyMatrices(XT, X);
  const XTY = multiplyMatrices(XT, Y);
  if (!XTX || !XTY) return null;

  const invXTX = invertMatrix(XTX);
  if (!invXTX) return null;

  const betaMatrix = multiplyMatrices(invXTX, XTY);
  if (!betaMatrix) return null;

  const coeffs = betaMatrix.map((row) => row[0]); // [a0, a1, a2, ...]

  const predict = (x: number) => {
    let y = 0;
    for (let j = 0; j <= degree; j++) {
      y += coeffs[j] * Math.pow(x, j);
    }
    return y;
  };

  // Residuals & R²
  const meanY = points.reduce((acc, p) => acc + p.y, 0) / n;
  let ssTot = 0;
  let ssRes = 0;
  const residuals: ResidualPoint[] = [];

  for (const p of points) {
    const yPred = predict(p.x);
    const res = p.y - yPred;
    ssTot += Math.pow(p.y - meanY, 2);
    ssRes += Math.pow(res, 2);
    residuals.push({
      x: p.x,
      yObserved: p.y,
      yPredicted: yPred,
      residual: res,
    });
  }

  const rSquared = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 1;
  const standardError = n > numCoeffs ? Math.sqrt(ssRes / (n - numCoeffs)) : 0;

  // Build equation string: y = a_n x^n + ... + a1 x + a0
  const terms: string[] = [];
  for (let j = degree; j >= 0; j--) {
    const c = coeffs[j];
    if (Math.abs(c) < 1e-8 && terms.length > 0) continue;
    if (j === 0) {
      terms.push(`${c >= 0 && terms.length > 0 ? '+ ' : ''}${c.toFixed(4)}`);
    } else if (j === 1) {
      terms.push(`${c >= 0 && terms.length > 0 ? '+ ' : ''}${c.toFixed(4)}x`);
    } else {
      terms.push(`${c >= 0 && terms.length > 0 ? '+ ' : ''}${c.toFixed(4)}x^${j}`);
    }
  }

  return {
    type: 'polynomial',
    degree,
    equation: `y = ${terms.join(' ')}`,
    r: Math.sqrt(rSquared),
    rSquared,
    coefficients: coeffs,
    predict,
    residuals,
    standardError,
  };
}

/**
 * Fits Exponential Regression: y = a * e^(bx)
 * By transforming to ln(y) = ln(a) + bx. Requires all y > 0.
 */
export function fitExponentialRegression(points: DataPoint[]): RegressionResult | null {
  if (points.length < 2 || points.some((p) => p.y <= 0)) return null;

  const transformed: DataPoint[] = points.map((p) => ({ x: p.x, y: Math.log(p.y) }));
  const lin = fitLinearRegression(transformed);
  if (!lin) return null;

  const b = lin.coefficients[1];
  const a = Math.exp(lin.coefficients[0]);

  const predict = (x: number) => a * Math.exp(b * x);

  const meanY = points.reduce((acc, p) => acc + p.y, 0) / points.length;
  let ssTot = 0;
  let ssRes = 0;
  const residuals: ResidualPoint[] = [];

  for (const p of points) {
    const yPred = predict(p.x);
    const res = p.y - yPred;
    ssTot += Math.pow(p.y - meanY, 2);
    ssRes += Math.pow(res, 2);
    residuals.push({ x: p.x, yObserved: p.y, yPredicted: yPred, residual: res });
  }

  const rSquared = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 1;

  return {
    type: 'exponential',
    equation: `y = ${a.toFixed(4)} · e^(${b.toFixed(4)}x)`,
    r: Math.sqrt(rSquared),
    rSquared,
    coefficients: [a, b],
    predict,
    residuals,
    standardError: Math.sqrt(ssRes / (points.length - 2)),
  };
}

/**
 * Fits Logarithmic Regression: y = a + b * ln(x)
 * Requires all x > 0.
 */
export function fitLogarithmicRegression(points: DataPoint[]): RegressionResult | null {
  if (points.length < 2 || points.some((p) => p.x <= 0)) return null;

  const transformed: DataPoint[] = points.map((p) => ({ x: Math.log(p.x), y: p.y }));
  const lin = fitLinearRegression(transformed);
  if (!lin) return null;

  const a = lin.coefficients[0];
  const b = lin.coefficients[1];

  const predict = (x: number) => (x > 0 ? a + b * Math.log(x) : NaN);

  const sign = b >= 0 ? '+' : '-';
  return {
    type: 'logarithmic',
    equation: `y = ${a.toFixed(4)} ${sign} ${Math.abs(b).toFixed(4)} · ln(x)`,
    r: lin.r,
    rSquared: lin.rSquared,
    coefficients: [a, b],
    predict,
    residuals: lin.residuals.map((r, i) => ({
      x: points[i].x,
      yObserved: points[i].y,
      yPredicted: r.yPredicted,
      residual: r.residual,
    })),
    standardError: lin.standardError,
  };
}

/**
 * Fits Power Regression: y = a * x^b
 * Requires all x > 0 and y > 0.
 */
export function fitPowerRegression(points: DataPoint[]): RegressionResult | null {
  if (points.length < 2 || points.some((p) => p.x <= 0 || p.y <= 0)) return null;

  const transformed: DataPoint[] = points.map((p) => ({
    x: Math.log(p.x),
    y: Math.log(p.y),
  }));
  const lin = fitLinearRegression(transformed);
  if (!lin) return null;

  const a = Math.exp(lin.coefficients[0]);
  const b = lin.coefficients[1];

  const predict = (x: number) => (x > 0 ? a * Math.pow(x, b) : NaN);

  const meanY = points.reduce((acc, p) => acc + p.y, 0) / points.length;
  let ssTot = 0;
  let ssRes = 0;
  const residuals: ResidualPoint[] = [];

  for (const p of points) {
    const yPred = predict(p.x);
    const res = p.y - yPred;
    ssTot += Math.pow(p.y - meanY, 2);
    ssRes += Math.pow(res, 2);
    residuals.push({ x: p.x, yObserved: p.y, yPredicted: yPred, residual: res });
  }

  const rSquared = ssTot > 0 ? Math.max(0, 1 - ssRes / ssTot) : 1;

  return {
    type: 'power',
    equation: `y = ${a.toFixed(4)} · x^(${b.toFixed(4)})`,
    r: Math.sqrt(rSquared),
    rSquared,
    coefficients: [a, b],
    predict,
    residuals,
    standardError: Math.sqrt(ssRes / (points.length - 2)),
  };
}
